### DISCUSSION TITLE
Execution Platform Scoped Spawn Strategies

### DISCUSSION THREAD

**Discussion by Silic0nS0ldier** [bazelbuild/bazel#20497] [Design proposals]
While Bazel has the necessary primitives to configure a multi-platform build, it currently lacks a way to map registered execution platforms to spawn strategies. This complicates remote execution adoption by forcing indirect adjustments to be made such as;

* Tagging targets/actions with `no-remote-exec` to prevent the `remote` strategy being used for select targets.
* Keeping local and remote spawned action graphs separate, and setting global configuration according to which graph a target belongs to in `bazel build`.
* Keeping the platform consistent across remote and local.
* Taking advantage of unique properties of actions that are intended to be spawned remotely, like their mnemonics.

To date there have been 2 proposals to address this problem.
- 2019-08-01 - [2019-06-25-platforms-strategies.md](https://github.com/bazelbuild/proposals/blob/1032d7d8fbe0e518e88482e7c532eb409ef6f1ac/designs/2019-06-25-platforms-strategies.md) by @katre
- 2023-03-31 - [Execution Platforms vs. Strategies](https://docs.google.com/document/d/1U9HzdDmtRnm244CaRM6JV-q2408mbNODAMewcGjnnbM) by @katre and @aragos

I've have a new one to add to the pile, though with a much reduced scope. The gist is that the existing spawn strategy flags (e.g. `--spawn_strategy`, `--strategy=<mnemonic>`) will be extended to accept an execution platform as a filter.

e.g.

```
# When the exec platform is "//:darwin_arm64" use one of "worker,sandboxed,local" by default
build --spawn_strategy=//:darwin_arm64=worker,sandboxed,local
```

This is intended to fill a capability gap while work on #11432 continues.

Proposal;
- 2023-12-07 - https://github.com/bazelbuild/proposals/blob/1423671b9176d057eda1a42e0e2a469c32ab2476/designs/2023-06-04-exec-platform-scoped-spawn-strategies.md
- 2024-02-10 - https://github.com/bazelbuild/proposals/blob/14f45b8ede1ba162ef7b2e7fdd26eda24409bd47/designs/2023-06-04-exec-platform-scoped-spawn-strategies.md

cc @katre

**Comment by katre** [comment]
Thanks for suggesting this! I have a lot of comments, but the top-level one is that I think this is a good step forward, and will be useful while we try to figure out a larger plan.

I'm going to split out comments into separate discussion threads, I hope this is easier, let me know if it isn't.

**Comment by katre** [comment]
> This does not seek to solve;

This proposal aims to select the correct strategy for a given execution platform. It is not trying to solve the problem of picking the execution platform based on the execution strategy (as in [Execution Platforms vs. Strategies](https://docs.google.com/document/d/1U9HzdDmtRnm244CaRM6JV-q2408mbNODAMewcGjnnbM). Can we make that more explicit?

**Reply by Silic0nS0ldier** [reply]
I've clarified this in https://github.com/bazelbuild/proposals/pull/363. If its too buried in the details let me know, might be worth stating this in the abstract.

**Comment by katre** [comment]
> RuleContext#getExecutionPlatform(), null if getToolchainContext() returns null

In theory, if `RuleContext#getToolchainContext()` returns null, it means that the rule isn't generating any actions. That said, we don't have any validation, so it's entirely possible that legacy rules still violate this. (I've filed https://github.com/bazelbuild/bazel/issues/20505 to try to address this).

So, given that, you should assume the worst: that some rule will generate a Spawn with no execution platform set that should actually try to run (somewhere).

**Reply by katre** [reply]
I've vetted this (see [my comment](https://github.com/bazelbuild/bazel/issues/20505#issuecomment-1856144402)), and it turns out to be fairly safe: only file write actions ever execute without an execution platform set (and based on my reading those are implicitly the host platform: while there is an [`isRemote` parameter in `FileWriteActionContext.writeOutputToFile`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/analysis/actions/FileWriteActionContext.java;drc=51f71f1a0f12a597e6166e43907e9f2d0de0222e;bpv=1;bpt=1;l=31?gsn=writeOutputToFile&gs=KYTHE%3A%2F%2Fkythe%3A%2F%2Fgithub.com%2Fbazelbuild%2Fbazel%3Flang%3Djava%3Fpath%3Dcom.google.devtools.build.lib.analysis.actions.FileWriteActionContext%23198318e8977f945737753f4accc1ae0f8272cf2534feeca4fbefa5888a5c4ae9), it's [ignored in Bazel's only implementation](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/exec/FileWriteStrategy.java;drc=51f71f1a0f12a597e6166e43907e9f2d0de0222e;l=49)).

However, there's no clear way to add a validation to make sure this remains the case, so we may need some degree of defensive coding.

**Reply by Silic0nS0ldier** [reply]
In https://github.com/bazelbuild/proposals/pull/363 I have updated the risk notes to account for the investigation results, and included potential mitigations (accept the risk and fix bugs if they come up, or refactor so `null` isn't a potential value for the exec platform).

**Comment by katre** [comment]
Here's the big one: I like the idea but dislike the flag syntax you've chosen.

I don't know how deeply you've looked into Bazel's flag parsing: it's very specific and arcane (and we'd love to rip it out and replace it wholesale, but [Hyrum's law](https://www.hyrumslaw.com/) applies: any replacement would have to be bug-for-bug compatible semantically and syntactically, and so isn't worth it).

> The likihood of misconfiguration can be reduced with;
> 
> --[no]require_platform_scoped_strategies to make inclusion of platforms in spawn strategy flags mandatory.
> --[no]exhaustive_platform_strategies to make specifiction of spawn strategies for every registered execution platform required.

I don't think these are possible, for a few reasons:
1. You can't access other flags while parsing a flag, so there's no easy way to tell, while dealing with `--spawn_strategy`, that `--require_platform_scoped_strategies` is enabled.
  a. This means you'd need to add a validation step after flags are parsing, and before any analysis starts, to ensure that the strategies are set correctly. It might be possible to do this, but there's not a single central place for it.
   b. See [`BuildCommand`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/runtime/commands/BuildCommand.java`), [`TestCommand`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/runtime/commands/TestCommand.java), and [`AqueryCommand`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/runtime/commands/AqueryCommand.java?q=AqueryCommand&ss=bazel) as three places that need this handling. All of them use [`BuildRequest`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/buildtool/BuildRequest.java), so maybe that would work, but I'm not sure we can guarantee it.
2. Furthermore, I'm not sure there's a single answer to "what are the available execution platforms?"
  a. The canonical answer is to fetch [`RegisteredExecutionPlatformsValue`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/skyframe/toolchains/RegisteredExecutionPlatformsValue.java), but the `Key` class in that depends on the current configuration to check the value of `--extra_execution_platforms`.
  b. You can't easily call into Skyframe from `BuildRequest`, because it's before Skyframe has initialized.
  c. While I'm not aware of any transitions that change `--extra_execution_platforms`, it's currently possible (thanks I hate it). I am aware of transitions while change the related `--extra_toolchains`.

> \# For actions with the "FOO" mnemonic and exec platform "//:darwin_arm64" use "local"
> build --strategy=FOO=//:darwin_arm64=local

The data structure for [`--strategy`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/exec/ExecutionOptions.java?q=symbol%3A%5Cbcom.google.devtools.build.lib.exec.ExecutionOptions.strategy%5Cb%20case%3Ayes) is the very complicated:

```
  @Option(
      name = "strategy",
      allowMultiple = true,
      converter = Converters.StringToStringListConverter.class,
      defaultValue = "null",
      ...)
  public List<Map.Entry<String, List<String>>> strategy;
```

The `List<Map.Entry` part is how Bazel's options parser handles key-value `allowMultiple` options: in theory this could eventually be collected to an actual `Map` but really it's directly iterated [in `BazelStrategyModule`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/bazel/rules/BazelStrategyModule.java;drc=51f71f1a0f12a597e6166e43907e9f2d0de0222e;l=86) to be added to a `SpawnStrategyRegistry`.

Making this more complicated with a "key1=key2=value" flag seems overly complicated: not so much semantically for users (although I think it is), but definitely as an implementation detail.

**Reply by katre** [reply]
To summarize, we need the following features:
1. We need to track strategies by mnemonic (the current `--strategy` flag), by description (the current `--strategy_regexp` flag), and as a default (the current `--spawn_strategy` flag).
2. We need to be able to parameterize the strategy for any of these by execution platform label.
3. We also need to set defaults for each to handle both "execution platform not specified" and "no execution platform at all".

**Reply by katre** [reply]
Checking in with @lberki for overall approach: Bazel would prefer to minimize/remove the use of existing flags, and avoid adding new ones, correct?

For @Silic0nS0ldier (and other interested parties): how often do users want to set strategies? Is this a common operation, or do the various strategy flags only exist in bazelrc files?

Finally, an actual suggestion: rather than use flags to configure this, what if we added a new attribute to execution platforms (similar to [the proposal I made in [`2023-06-04-exec-platform-scoped-spawn-strategies.md`](https://github.com/bazelbuild/proposals/blob/1032d7d8fbe0e518e88482e7c532eb409ef6f1ac/designs/2019-06-25-platforms-strategies.md) to track what strategies to use with a specific execution platform? This would be more limited than that proposal (it only affects strategy selection).

It would also be more limited than the current proposal, in that it only allows setting the strategy based on the execution platform, not the execution platform and the mnemonic (or description). Is this actually needed, or was it only included for convenience?

I await your response to my strawman counter-proposal.

**Reply by lberki** [reply]
Reducing the number of flags is only a proxy goal. The actual goal is to reduce the complexity of the interface of Bazel in check, it's just that the number of command line flags is a reasonable proxy for that, I think.

In this particular case, I think the decision should be made based on who knows what to do: if it's the person who defines the execution platform who knows which strategies work on it, it should be part of the platform definition, if it's the end user (or their tech lead, etc.), it should be a command line flag.

Without knowing the details, my intuition would be that this should be a command line flag, simply because the way we tell Bazel what actions to run with which strategy is currently wholly specified on the command line. As you can see, this is not a very sophisticated reasoning :)

**Reply by katre** [reply]
>  if it's the person who defines the execution platform who knows which strategies work on it, it should be part of the platform definition

I actually think that is the case: it also helps to explain why it'd be useful to define multiple similar platforms. For example, the automatic `@local_config_platform//:host` platform would obviously support "local" (and probably "worker" and "sandboxed"). If you're using a remote execution system, then the platforms that defines would only use "remote".

This is the impetus for the original desire for strategies driving execution platform selection (it lets us go from "strategy=remote" to "use `@remote//:platform`"). This is especially important since I think all remote execution systems need configuration information from the execution platform: if a platform with identical constraints but different exec properties is chosen the remote system might still be used but will be set up wrong.

Given that this is already a problem, I think the current proposal is fine: users need to tune the order of their execution platforms to prefer the remote execution platform **now**, so needing to do that in the future doesn't make more work, and this proposal makes sure that Bazel doesn't try to use the remote execution platform with a local strategy.

**Reply by Silic0nS0ldier** [reply]
I've yet to fully review all the comments here, but regarding defining compatible spawn strategies on platforms (an approach I explored before landing on the current proposal) here are my thoughts;

* `platform` declarations do not always live in the workspace (main repo) nor is `platform(...) necessarily directly called. They may be generated by a macro, repository rule, extension, or be a static part of a ruleset. This limits their usefulness for expressing spawn strategy support.
* The current proposal limits scope to exclusively the main repo (I'm ignoring that transitions can change flags, I haven't seen spawn strategy flags used in that context and frankly I hope I never do). The first point notes that `platform` declarations can come from third party sources, so a solution that involves changing that API has a bigger blast radius.
  As an example, if/when a more comprehensive solution lands the existing API will need to be removed. That means any ruleset which starts using it is going to later need to be updated to remove usage to be compatible with newer versions of Bazel.
* Platform inheritance would need to be considered in the design (more scope).
* This is more of a passing thought, but I'm not sure it makes sense for platforms themselves to have _any_ execution related information.
  Consider the existing `exec_properties` attribute. Bazel (according to documentation) does nothing with this information, most usage I see relates to configuring remote execution. It is only useful for platforms used as an execution platform, a target platform has no use for it. Additionally the kind of data set via this attribute is more closely aligned with spawn strategies. It may make sense to restrict platforms to exclusively a configuration role, with a separate primitive used to inform if and how a platform can be used for execution. From a certain point of view, that is exactly what this proposal does. It just lacks the ability to tie in data like you'd see in `exec_properties`. An idea close to this was explored in [Execution Platforms vs. Strategies 2023-03-31](https://docs.google.com/document/d/1U9HzdDmtRnm244CaRM6JV-q2408mbNODAMewcGjnnbM) with `*_strategy_settings(...)`.
* Less important but worth noting, encoding spawn strategy compatibility in platforms vs. scoping spawn strategies by execution platforms are not functionality equivalent. The proposal as is allows more combinations to be expressed (for better or worse).

Oh, and regarding the flag syntax. I'm not a fan myself either. The suggestion about encoding compatibility in platforms lends itself to a different flag-oriented idea though. We could simply add a new flag in which the exact same information can be encoded.

e.g. (imagine a better argument name, less prone by semantic satiation induced by the many other strategy flags)
```
--execution_platform_strategies=//:foo=worker,local
--execution_platform_strategies=//:bar=remote
```

**Reply by fmeum** [reply]
> platform declarations do not always live in the workspace (main repo) nor is `platform(...) necessarily directly called. They may be generated by a macro, repository rule, extension, or be a static part of a ruleset. 

While I do know of a number of examples and rules_go itself is guilty of it, I've always wondered whether this isn't fundamentally wrong: A `platform` ties in together all the different dimensions that matter in a build and no single ruleset can be aware of all of them. I'm starting to think that any kind of hard coded platform in a ruleset or external repo that isn't meant to be used solely as a parent platform is conceptually problematic.

Maybe one part of the solution to this problem space is to pin down best practices on how platforms should "flow through" builds?

**Reply by katre** [reply]
Ideally, rules repos won't define execution platforms, but configuration repos for remote execution might. (Rules repos defining **target** platforms is also worrying to me, but it's a different problem and hits issues of user experience that I don't have good answers for).

That said, your point is valid: it's worth not expanding the surface of the `platform` rule. At one point we discussed a separate `execution_platform` rule, and that would be a valid place to declare compatible strategies, but we're not there now.

I like the `--execution_platform_strategy` flag (note change in plurality) as a concept, but we need to clearly define the interaction when I have:
1. An action with description "Frobbing the quzzle component", mnemonic "FROB", and execution platform "//platform:macos_arm_remote_cluster"
2. `--strategy_regexp=Frobbing.*=local`
3. `--strategy=FROB=sandbox`
4. `--execution_platform_strategies=//platform:macos_arm_remote_cluster=remote`

I _think_ in this case the execution platform strategy should win, but I'm not 100% on that.

**Reply by Silic0nS0ldier** [reply]
I think I miscommunicated the intent behind the `--execution_platform_strateg[ies|y]` flag. It was late for me when I sent the message, so I'm not surprised.

Rather than being another way to specify the spawn strategy to use, the idea was to declare the spawn strategies the execution platform supports. Such that we can state an exec platform supports for example `worker,sandboxed,local` to exclude `remote` as a possible spawn strategy. The placeholder name probably didn't help.

**Reply by katre** [reply]
Ahh, I see. The similarity to the other strategy flags confused me, but that makes much more sense.

Can you update the proposal with these ideas, and then we'll do a second round?

**Reply by Silic0nS0ldier** [reply]
PR is up for revised proposal. https://github.com/bazelbuild/proposals/pull/363

**Comment by katre** [comment]
Thanks very much for the updates.

My only remaining question is that I'm not entirely sure how `--require_platform_scoped_strategies` would be implemented: have you begun to look into coding any of this?

Aside from that (and possibly seeking any idea for flag names: `--execution_platform_supported_strategies` works but is a bit long) I'm happy to approve this. Thanks for bringing this proposal to the project.

Let me know what your timeline looks like for implementation. I'm happy to review PRs and help get them imported, although because of the holidays today is my last working day until January 2, 2024.

**Reply by Silic0nS0ldier** [reply]
@katre I created a prototype for this proposal in its initial form. https://github.com/Silic0nS0ldier/bazel/pull/5

I'm thinking of revising the prototype by;
* Using a spawn strategy wrapper (like `PlatformScopedSpawnStrategy`) that rejects a list of exec platforms.
* In `SpawnStrategyRegistry.toStrategy(...)` when the new flag is used, wrapping existing spawn strategies. Here is where the rejection lists will be built.

As for enforcing that all execution platforms must specify their supported strategies, I have an idea for a naive implementation. During spawn strategy resolution a check against the flag can be performed. If the current execution platform isn't present an error can be raised. This is not optimal (error will not always show up and adds to the cost of each action spawn), my aim will be to implement this check early on (immediately after all execution platforms have been registered).

As far as timelines for the implementation go, it'll be this quarter. 🤞 in February.

I'll do some brainstorming on flag names and then get the proposal updated.

**Reply by katre** [reply]
Your prototype looks sound, I don't know if you want me to comment on it yet. I like the idea of the `PlatformScopedSpawnStrategy`, I think that part makes a lot of sense.

Looking forward to the proposal updates, I think this is basically ready to go ahead. Thanks so much!

**Reply by Silic0nS0ldier** [reply]
Proposal update is up. https://github.com/bazelbuild/proposals/pull/367

> Your prototype looks sound, I don't know if you want me to comment on it yet.
The linked prototype is for the original proposal, and mainly served as a PoC. It illustrates my current thinking on how to approach the implementation, but commenting/reviewing wouldn't be the best use of our time.

When a proper implementation is ready, I'll reach out for comments then.

**Reply by Silic0nS0ldier** [reply]
I've got an incomplete prototype up for the revised proposal. If I'm misusing, abusing, or otherwise putting things in the wrong place let me know and I can course correct.
https://github.com/Silic0nS0ldier/bazel/pull/6

**Reply by katre** [reply]
Thanks, this looks fine to me. I have a small suggestion for the flag name, but otherwise I'm happy to approve this proposal.

**Reply by Silic0nS0ldier** [reply]
Flag name tweaked in https://github.com/bazelbuild/proposals/pull/367.

I'll give it a final (🤞) look tomorrow and put it up for the approval review.

**Comment by katre** [comment]
Hi, I'm trying to follow up on this. I'm still happy with the proposal, but I never saw a PR to mark it as approved, and I haven't seen any incoming implementation PRs. Is this still something you're interested in working on?

(In a related area, see my [proposal for Selectively Enabling Execution Platforms](https://github.com/bazelbuild/bazel/discussions/22170)).

**Reply by Silic0nS0ldier** [reply]
Hi @katre, I haven't opened a WIP PR against the Bazel repo but do have https://github.com/Silic0nS0ldier/bazel/pull/5. I am still interested in working on this, I just haven't had sufficient time to focus on pushing this over the line for the past few weeks. My schedule has since cleared thankfully. With luck I'll have a review-appropriate prototype up this weekend.

From my cursory glance the linked proposal looks useful. I'll have a more thorough look at it next week and give my thoughts.

**Reply by katre** [reply]
Glad to hear it, I was checking on old projects and realized I didn't know what the status on this one is.

Feel free to send a PR to proposals to mark this as an accepted proposal, just so that the metadata is up to date. I'll look forward to seeing your code once you're ready for a review.

**Reply by Silic0nS0ldier** [reply]
A quick update. The approach I was taking in https://github.com/Silic0nS0ldier/bazel/pull/6 had some holes so I've been reevaluating, which is actually what has had me holding off on pushing the proposal into "accepted".

To better equip myself for the next attempt I've spent some time getting a [more complete development environment](https://github.com/Silic0nS0ldier/bazel-vscode-env) together. I gave it a road test in
- #22716
- #22742
- #22762
All small, (mostly) inconsequential changes to help work out the kinks (there [were](https://github.com/georgewfraser/java-language-server/pull/288) [some](https://github.com/georgewfraser/java-language-server/pull/289)).

I also put it through its paces investigating an elusive server process hang as well, the one mentioned in 
https://github.com/bazelbuild/bazel/blob/3f93d377d036d773fd505a18e084425a00fb94ea/src/main/java/com/google/devtools/build/lib/runtime/BlazeCommandDispatcher.java#L190-L191
I'll be submitting an issue with a repro for that ASAP (whatever that counts for when a custom client process is necessary to increase the incidence rate enough).

Going back on topic... I'll be restarting the implementation effort from now. I'll likely approach this from the bottom-up (i.e. starting from the strategy classes), so there is less chance that I'll misidentify similarly named symbols.

**Reply by Silic0nS0ldier** [reply]
First sliver of (hopefully merge-able) work. #22925

Next up, trying to make `getExecutionPlatform()` never return `null`. From my skim `null` can arise in a few scenarios currently.
1. Builtin actions like `MiddlemanAction`, `SymlinkAction` and `SolibSymlinkAction`.
2. `RuleContext` where exec platform is resolved from toolchain context, which may be `null`.
3. Exec groups (I think, cannot track down where right now).

(1) isn't too much of a worry. A "bazel internal" platform stub could be used satisfy non-null requirements. (2) may be a non-issue since the class in question states instances only exist during analysis. (3) seems like it would represent a bug (if it were to ever happen).

There are several code comments stating that `null` for `PlatformInfo` means the host platform should be assumed. Given the latter context, this may actually be referring to builtin actions which are unaffected by spawn strategy controls. My current thinking is that this unique scenario may benefit from a more explicit representation, to distinguish between bugs (`null` exec platform) and builtin actions.

**Reply by katre** [reply]
Thanks for the update.

Despite being represented as an action, `MiddlemanAction` isn't actually executable, and so a null execution platform is reasonable. On the other hand, `SymlinkAction` and `SolibSymlinkAction` are actually executable and should have real execution platforms.

`RuleContext` and `ExecGroup` actually use the same underlying logic (all of the `RuleContext` toolchain-related code delegates to the default `ExecGroup`), which means that the execution platform cannot actually be null if the rule can create actions (some rules, like [`PlatformRule`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/rules/platform/PlatformRule.java;drc=b5b5ca1399ff1fa60338449e7c5b6327bb280f69;l=52) explicitly disable toolchain resolution and should never create actions). We don't have good checks for this: it'd be nice to add a check that these rules **don't** actually create actions, but the implicit expectation is that any rule which can create actions has an execution platform. Cleaning all of this up definitely has value.

Do you intend any further changes to this proposal? Feel free to send me PRs but I'd like to get it marked as approved, even if it will change during implementation, just so that we know the current state. Then you can start sending out implementation PRs.

**Reply by Silic0nS0ldier** [reply]
PR to put proposal into approved is up. https://github.com/bazelbuild/proposals/pull/387

**Reply by Silic0nS0ldier** [reply]
Been a lot of progress around this since July, if not on this specifically.

- [Java Language Support](https://github.com/georgewfraser/java-language-server) was forked as [Java+Bazel Language Support](https://github.com/Silic0nS0ldier/java-language-server). As of [`v0.3.1`](https://github.com/Silic0nS0ldier/java-language-server/releases/tag/v0.3.1);
  - Import resolution is fully operational in the Bazel repo. Previously external dependencies (e.g. Google base libraries, JavaX, JUnit) and generated code (protos) would not resolve.
  - No longer crashes with `Compiler is already in-use` after a few minutes of usage.
  - Reference lookups such as "Find All References" now consistently work, no longer crashing with an out of bounds error.
  - Updated workspace detection to support bzlmod changes and handle extensioned (`.bazel`) versions.
- The [dev container for Bazel](https://github.com/Silic0nS0ldier/bazel-vscode-env) has been refined through usage via Bazel migration projects ([`Silic0nS0ldier/java-language-server`](https://github.com/Silic0nS0ldier/java-language-server/tree/5f06ae856bab034354c399200456f154247c1421/.devcontainer) and [`Silic0nS0ldier/vscode-git-monolithic-extension`](https://github.com/Silic0nS0ldier/vscode-git-monolithic-extension/tree/b45479e545ab4c14a095ff7e7edd445629b7fb4a/.devcontainer)).
- Practise, practise, practise.
  - (merged) #23233
  - (merged) #23375
  - (discarded) #23252
  - (discarded) #23374

Progress on this...
- #23231
  This will help to cut down on places where `null` could creep in. Formally at least, annotations don't count for much without something to validate (e.g. [NullAway](https://github.com/uber/NullAway), [JSpecify](https://jspecify.dev/), [Infer - Nullsafe](https://github.com/facebook/infer/blob/main/infer/annotations/src/main/java/com/facebook/infer/annotation/Nullsafe.java)).
- https://github.com/Silic0nS0ldier/bazel/pull/6
  I'm looking over this, and attempting to recreate the prototype. So far...
  - With the help of find-all-references, I tracked down where `SpawnStrategyRegistry` is constructed (in `src/main/java/com/google/devtools/build/lib/buildtool/ExecutionTool.java`).
    CLI options can be access in this scope via `env` (`CommandEnvironment`).
    e.g. `env.getOptions().getOptions(ClientOptions.class)`
  - At that same site I can access the product name (possibly as many as 3 ways, all via `env`).
    With this I can implement a `Precondition` check to ensure in Bazel, anything asking for a spawn strategy must have an execution platform. This is to favour correctness over a defect (null platform) allowing the wrong strategy to be selected.

I'm aiming to have this implemented to an experimental level ahead of the Bazel v8 release branch being cut (this month).

**Reply by Silic0nS0ldier** [reply]
Took longer than planned (needed a solid block of focus time to hammer this out) but I've finally got it done!

https://github.com/bazelbuild/bazel/pull/24265

Half done... The misconfiguration safeguard `--require_platform_scoped_strategies` needs some more work. Had some issues threading through the CLI options into [`src/main/java/com/google/devtools/build/lib/skyframe/toolchains/RegisteredExecutionPlatformsFunction.java`](https://github.com/Silic0nS0ldier/bazel/blob/17135c347f77d852d7dff0e3802ca1b1cdee9775/src/main/java/com/google/devtools/build/lib/skyframe/toolchains/RegisteredExecutionPlatformsFunction.java).

**Comment by katre** [comment]
@Silic0nS0ldier: Any updates on the status for this? This is my last week at Google, so I won't be able to help review or merge in the future. Further reviews will be handled by @gregestren. Good luck!
